### PR TITLE
feat: add binary search

### PR DESCRIPTION
**Pull request by garrett361**
## Description
Adds a very simple binary search. 

Randomly generates as many configurations as can be concurrently runs binary search on all of them. Either a binary search completes before `max_trials` is hit, in which case a new conig is generated and BS is started over again for that config, or the search is interrupted when `max_trials` is hit.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/tests/experiment/pytorch/test_deepspeed_autotuning.py [hunk 1]
```diff
@@ -47,11 +47,11 @@
...
-    "num_params": 60192808,
-    "trainable_num_params": 60192808,
-    "activation_mem_per_gpu": 89828352,
+    "num_params": 1,
+    "trainable_num_params": 1,
+    "activation_mem_per_gpu": 1,
     "rank": 0,
-    "gpu_mem": 15843721216,
+    "gpu_mem": 32,
 }
```
Comment (tayritenour) [inline harness/tests/experiment/pytorch/test_deepspeed_autotuning.py:55]: Why would we change this? wouldn't we want values that are reasonably realistic?

**Review by tayritenour — APPROVED** [review]
LGTM

File: harness/tests/experiment/pytorch/test_deepspeed_autotuning.py (hunk 1 above)
Comment (garrett361) [reply harness/tests/experiment/pytorch/test_deepspeed_autotuning.py:55]: It made binary search tests run for an annoyingly long time. Like ~300 different checks per stage. These settings dropped that down to ~30.  Can change it back, if you prefer, though.

Comment (tayritenour) [reply harness/tests/experiment/pytorch/test_deepspeed_autotuning.py:55]: I suppose it's fine. I just want to make sure our tests are reasonably representative

**Pull request merged**